### PR TITLE
architectures folder for easier porting to embedded devices

### PR DESCRIPTION
**Pull request by Pro**
We started with the architecture folder implementation and this PR is open for discussion. It is still WIP since it needs some further cleanup, but the majority of the functionality should be there.

The idea is:
in `plugins/arch` we have a folder for every architecture which contains any OS specific code. Any code outside the arch folder should solely rely on the defined functions from the arch folder. 
When compiling for a specific architecture (see the `UA_ARCHITECTURE` CMake Variable), the corresponding `ua_architecture.h` file will be included and with this you can easily switch between architectures.
E.g. see `ua_network_tcp.c` for an example.

The list of available architectures is automatically extended by CMake if someone adds another architecture folder.

This will also allow easy integration of new hardware platforms and allows commercial users to implement their own architecture driver without the requirement of distributing it under MPL 2.0, which may be desirable for some.

### CODE REVIEW

**Comment by Pro** [comment]
@open62541/open62541-core please look at the suggested changes and leave your comments here.

**Comment by jpfr** [comment]
Hey @Pro @cabralfortiss ,

this is cool!
I'd like to note that some "plugins" still are available for several platforms.
For example the nodestore and SecurityPolicies.

I'd separate based on whether the feature is baked into the SDK (e.g. the clock access) or whether it shall be added in the server/client configuration.

- /src (main library)
- /arch (arch-specific features to be compiled into the library, e.g. networking, clock, ...)
- /plugins (features to be added in the configuration, e.g. nodestore, encryption, ...)

Since we try to establish a common basis for the networking layer on all platforms, can we move the network layer to arch? So we can compile in a fixed network layer and don't add them dynamically in the config. The common networking code then be moved to /src/ua_connection.c or similar.

Also, does the arch code have the "right" to access internal API? The plugins currently don't.
This decision should be documented.

**Comment by Stasik0** [comment]
http://thinkrelevance.com/blog/2011/11/15/documenting-architecture-decisions

**Comment by Pro** [comment]
@open62541/open62541-core from our side this PR would be ready to be squashed and merged. If there aren't any objections, we can even target it for 0.3 release?!

**Comment by StalderT** [comment]
I tested the PR. As I understand it, using UA_ENABLE_AMALGAMATION, we need to have an open62541.c/h file for each platform? It will no longer be possible to have a single protable file?

**Comment by cabralfortiss** [comment]
That's right about each amalgamated file for each platform. Should we try to amalgamate all the available architecture code and put them behind a #ifdef so compiling will only need a variable. This will increase a little bit the size of the open62541.c/h of course

**Comment by Pro** [comment]
Yes, that's correct, I didn't think about that aspect...
As @cabralfortiss already suggested, we could include multiple architectures in the amalgamated file and surround them by an ifdef (i.e. just write `#ifdef UA_ARCHITECTURE_WIN32` in all the files within an architecture folder) and add all of them to the amalgamation. To compile for a specific architecture using amalgamation someone would then need to use a make flag like `-DUA_ARCHITECTURE_WIN32`

The question here is: which architectures do we want to include? Should we only include the most used ones (i.e. win32, posix) and for the others someone needs to download the source, or do we want to include all possible architectures in the arch folder.

**Comment by jpfr** [comment]
> The question here is: which architectures do we want to include? Should we only include the most used ones (i.e. win32, posix) and for the others someone needs to download the source, or do we want to include all possible architectures in the arch folder.

👍

**Comment by jpfr** [comment]
This is a big and important change. I would wait for 0.4 to get it right.
It might take a few weeks to erase all the little issues in the exotic architectures.

This is cool and already looking quite mature. I have only one request.
**Can we make the arch folder happen without introducing additional de-facto public interfaces?**
Changing public interfaces is possible between major releases. But it should be a separate discussion.

Right now, we force the implementation details of the TCP/POSIX networking into all architectures.
Basically we just distribute the #ifdef stuff over many files.
See also the Readme.md file.

But there are a lot of architectures that don't have getsockopt, getnameinfo and friends.
Also on Linux, there are cases where users want to be as close to the network device as possible and don't use BSD sockets.
See these projects:
http://dpdk.org/
https://github.com/AVnu/OpenAvnu

We have a plugin interface for every plugin. Also for networking.
Implementation details that are not part of the plugin interface should be exactly that: an implementation detail that can be changed in the background.
It is totally okay that several ports reuse the common parts of the POSIX code.
But it should not be mandatory to use POSIX concepts for all architecture ports.

On a lesser note, ua_malloc and friends are defined in /include/ua_config.h.in.
You include that at the very top. So no need to redefine for the architectures.

File: arch/freertos/ua_architecture.h [hunk 1]
```diff
@@ -0,0 +1,96 @@
+/* This work is licensed under a Creative Commons CCZero 1.0 Universal License.
+ * See http://creativecommons.org/publicdomain/zero/1.0/ for more information.
+ *
+ *    Copyright 2016-2017 (c) Julius Pfrommer, Fraunhofer IOSB
+ *    Copyright 2017 (c) Stefan Profanter, fortiss GmbH
```
Comment (HolgerJeromin) [inline arch/freertos/ua_architecture.h]: We are in 2018 right now :-)

Comment (Pro) [reply arch/freertos/ua_architecture.h]: Ah, thanks for the hint, my head was still in 2017, thanks to Copy & Paste 😄 
We also need to add @cabralfortiss to the list. But there is still some WIP for this PR.

**Comment by Pro** [comment]
It should now be possible to include multiple architectures in amalgamated files. See my latest commits.
If amalgamation is enabled but no specific architectures defined (`UA_AMALGAMATION_ARCHITECUTRES`) posix and win32 is added by default. You then need to define one of the two preprocessor defines to enable the corresponding architecture: `UA_ARCHITECTURE_POSIX` or `UA_ARCHITECTURE_WIN32`

Regarding the additional de-facto interfaces: IMHO the current state does not introduce new interfaces, but just cleans up the old ones. Up to now it is definitely an improvement to the state before. Supporting non-POSIX architectures would be a nice feature, but the question here is if we want to cover all of that in the scope of this PR, of if we want to merge this one first (after 0.3) and then see how we can get non-POSIX support. This would keep the PRs small. This one is quite big already.

**Comment by cabralfortiss** [comment]
I think the PR is in a mature state now. During this PR two more architectures were added (eCos and arduino). I think the most important imporovement with this PR is that a new architecture can be added completely independent from the existing code. 
I agree that this doesn't add anything new but just take everything arch-dependant out from the core code and put it in the arch folder.  
The things that are still missing to be done are:

- ua_config.h file which could be left as a general compiler/architecture header file (although I see that the first part of it, before #include "ua_architecture.h" can be in its own file, so it can be amalgameted first, so ua_architecture.h can use this variables)

- ua_network_tcp.c: has some arch-dependant code in the clientConnection to check if the connection was stablished

- ua_network_udp.c: completely to be reworked, similar to tcp

- UA_ENABLE_MULTITHREADING which uses mostly mutexes and some unkown-to-me function about interlocking

- Squash all the commits into one to have a better picture of what has been changed

**Comment by Pro** [comment]
The main functionality is now included in this PR.

The possibility to use non-POSIX architectures is currently not implemented and may require some additional work. I would prefer if we merge this PR first and then think of additional steps on how to add posix independability. That would mean that we would need to rewrite the ua_network_tcp (and the corresponding files for Pub/Sub). Before opening that box, we should first focus on other more important features/bugfixes.

@jpfr are you ok to merge? It still needs some minor tweaks, but we can do that after

**Comment by jpfr** [comment]
For master but not 0.3?
I would like to bring the Pub/Sub networking layer in first.
We have a bunch of additional PRs lined up that depend on Pub/Sub networking.

Is it okay if we aim for review+merge in a week or so?

**Comment by cabralfortiss** [comment]
Just some thoughts after reading people's usage of this in #1406.

Since the "arduino" and "freeRTOS + LwIP" platforms both use LwIP the networking part should be the same. In my case, memory allocation was done using the heap_4.c (https://www.freertos.org/a00111.html) implementation from freeRTOS since I couldn't use the normal malloc, free and so on. Since these platform were develop in parallalel maybe is time to merge them. 

I'd say to create a folder for common files inside the arch folder, and there have a lwip.h file with all the needed defines and everything from the networking, and another freeRTOS.h with all the freeRTOS functions. In this case, we should put UA_free and similars behind a define in order to have both of them, normal free and vPortFree. Maybe at the end will have only one platform. 

A problem I found is that I added my compiler flags to the CMakeLists.txt. Probably this should be left out, since other compilers will probably fail, and just use the CMAKE_C_FLAGS for this, but in this case the user should know the needed flags to compile.

**Comment by Grtschnk** [comment]
Adding my experience here with building/compiling open62541 for ESP32 (also mentioned it in [#1406](https://github.com/open62541/open62541/issues/1406)):

Branch: https://github.com/cabralfortiss/open62541/tree/feature/architectures
Board: Sparkfun ESP32,
Firmware: ESP-IDF,
IDE: Visual Studio 2017 + VisualGDB

"arduino" ua_architecture - got it to work (compile, run and succesfully connect):
1. #define UA_ARCHITECTURE_ARDUINO seems to be missing. Quickly fixed
2. 'access' does not seem to implement in ESP-IDF. I found in the header files but the linker(?) still can't find it. Added my own declaration that just returns zero.:
image
```
------ Build started: Project: Target, Configuration: Debug VisualGDB ------
VisualGDB: Run "D:\SysGCC\esp32\bin\bash.exe --login -c "export IDF_PATH=/esp-idf/v3.0 && export BATCH_BUILD=1 && export LANG=en_US && cd /cygdrive/c/Users/dtom077/Source/Repos/Kapua_1_3/Kapua_1_3 && make -j4 BUILD_DIR_BASE='/cygdrive/c/Users/dtom077/Source/Repos/Kapua_1_3/Kapua_1_3/Debug' SDKCONFIG='/cygdrive/c/Users/dtom077/Source/Repos/Kapua_1_3/Kapua_1_3/sdkconfig-debug' V=0"" in directory "" on local computer
including /esp-idf/v3.0/components/bootloader/Makefile.projbuild...
including /esp-idf/v3.0/components/bootloader_support/Makefile.projbuild...
including /esp-idf/v3.0/components/coap/Makefile.projbuild...
including /esp-idf/v3.0/components/esp32/Makefile.projbuild...
including /esp-idf/v3.0/components/esptool_py/Makefile.projbuild...
including /esp-idf/v3.0/components/mbedtls/Makefile.projbuild...
including /esp-idf/v3.0/components/nghttp/Makefile.projbuild...
including /esp-idf/v3.0/components/partition_table/Makefile.projbuild...
including /esp-idf/v3.0/components/ulp/Makefile.projbuild...
CC Debug/dccs/open62541.o
CXX Debug/main/main.o
including /esp-idf/v3.0/components/esptool_py/Makefile.projbuild...

[...]
                              ^
AR Debug/main/libmain.a
AR Debug/dccs/libdccs.a
LD Debug/Target.elf
/cygdrive/c/Users/dtom077/Source/Repos/Kapua_1_3/Kapua_1_3/Debug/dccs/libdccs.a(open62541.o)(0): error : undefined reference to `access'
/cygdrive/c/Users/dtom077/Source/Repos/Kapua_1_3/Kapua_1_3/Debug/dccs/libdccs.a(open62541.o): In function `UA_Discovery_cleanupTimedOut':
c:\Users\dtom077\Source\Repos\Kapua_1_3\Kapua_1_3\components\dccs\.\open62541.c(42228): error : undefined reference to `access'
/cygdrive/c/Users/dtom077/Source/Repos/Kapua_1_3/Kapua_1_3/Debug/dccs/libdccs.a(open62541.o): In function `process_RegisterServer':
c:\Users\dtom077\Source\Repos\Kapua_1_3\Kapua_1_3\components\dccs\.\open62541.c(42228): error : undefined reference to `access'
collect2: error: ld returned 1 exit status
make: *** [/esp-idf/v3.0/make/project.mk:389: /cygdrive/c/Users/dtom077/Source/Repos/Kapua_1_3/Kapua_1_3/Debug/Target.elf] Error 1
-------------------------------------------------------------
Command exited with code 2
Executable: D:\SysGCC\esp32\bin\bash.exe
Arguments: --login -c "export IDF_PATH=/esp-idf/v3.0 && export BATCH_BUILD=1 && export LANG=en_US && cd /cygdrive/c/Users/dtom077/Source/Repos/Kapua_1_3/Kapua_1_3 && make -j4 BUILD_DIR_BASE='/cygdrive/c/Users/dtom077/Source/Repos/Kapua_1_3/Kapua_1_3/Debug' SDKCONFIG='/cygdrive/c/Users/dtom077/Source/Repos/Kapua_1_3/Kapua_1_3/sdkconfig-debug' V=0"
Directory: 
VisualGDB: Error: Command-line action failed
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```

"freertos+lwip" ua_architecture (gave up at becuase of missing declaration in firmware, see 3.):
1. had to prepend 'freertos/' to the include path for the freertos files (task.h). Makes sense as ESP-IDF puts them there
2. UA_FREERTOS_HOSTNAME needs to be set. Okay.
3.  pvPortCalloc & pvPortRealloc don't seem to exist in ESP-IDF. Did not pursue further:
image

**Comment by Grtschnk** [comment]
And of course: Thank you, everyone, for creating/contributing to open62541!

**Comment by Grtschnk** [comment]
@cabralfortiss Successfully compiling and running your version of ua_architecture now as well!
I edited my previous post to add the information.

**Comment by Pro** [comment]
@andreasebner @jpfr is there an ETA for the last PubSub Batch? We agreed to wait until PubSub is integrated and then update this PR for that changes.

I get quite some requests to integrate this PR into master as soon as possible, since many users of open62541 want to use the stack for some proprietary hardware.

Are the coming changes for PubSub only on higher level or are there still changes in the network layer?
I guess that we mainly need to update the UDP network layer. The rest of the code should be architecture independent. Would it make sense to do that now and integrate this architecture PR even if the last PubSub batch is not yet integrated?

**Comment by Pro** [comment]
@jpfr are you ok to merge this as it is now? There are probably some minor tweaks necessary, but I would like to merge this and then create a new PR for these, since we constantly have merge conflicts here.

**Comment by jpfr** [comment]
Ehm. A squash-merge woulda been cool.
On github, that's one button click away from the regular merge.

Now we have 31 commits in our history from this. Most of them "fix", "cleanup" and "merge master".
That's just noise nobody needs.

https://github.com/open62541/open62541/blob/master/CONTRIBUTING.md#commit-and-pr-hygiene

**Comment by jpfr** [comment]
Anyway, cool to see this in!
Now we have a nice way to manage ports. :smile:

**Comment by lsstvrs** [comment]
@Pro 
Hi, I tried to compile the open62541 library to freertosLWIP but it fails as it lacks lwip related libraries ... should it be included externally? wouldn't it be easier to have them under deps folder?
Thanks

**Comment by cabralfortiss** [comment]
Normally you compile open62541 as a library so it shouldn't give any error on lacking of other libraries. The header files from LWIP and freeRTOS should be present in your computer, and you should point to them using the right variable in CMake. Both are open source project, so you can just clone the repos. 

Having them as deps could be too intrusive to this project, IMHO, since the architecture is not on the top list of used architectures.

**Comment by lsstvrs** [comment]
Alright... I will try pointing them on CMake... will let you know if I have any problem.

**Comment by FishEatingCat** [comment]
emmm......I don't know how to compile open62541 for freertos. I didn't see the compilation of freertos in the documentation of open62541. Can someone tell me how to compile open62541 for freertos? Thank you very much!

**Comment by cabralfortiss** [comment]
> 
> 
> Normally you compile open62541 as a library so it shouldn't give any error on lacking of other libraries. The header files from LWIP and freeRTOS should be present in your computer, and you should point to them using the right variable in CMake. Both are open source project, so you can just clone the repos.
> 
> Having them as deps could be too intrusive to this project, IMHO, since the architecture is not on the top list of used architectures.


As here said, you need to clone the repo of LWIP and of freeRTOS, both are C code. Then in open62541, checkout the master branch, and in CMake, choose freeRTOSLWIP as UA_ARCHITECTURE. You will get then more options. The most important are: 

- UA_ARCH_EXTRA_INCLUDES which you should point to the folders where the headers files from LWIP and freeRTOS, separeted by semicolon (;). 

- UA_ARCH_ADD_FLAGS, UA_ARCH_LINKER_FLAGS and UA_ARCH_REMOVE_FLAGS must be set according to the compiler you're using, so I cannot give more details now here. What I did was: in my IDE where I was compiling my normal C code for freeRTOS, I copy the flags and used them. Then when compilation fails, I had to remove some compilers flags (-Wconversion in my case) which weren't supported or were causing compilation errors. 

- UA_ARCH_FREERTOS_USE_OWN_MEMORY_FUNCTIONS is a boolean that you can set to true if you don't want to use free/malloc/calloc/realloc but the ones from the external implementation of freeRTOS called vPortFree/pvPortMalloc/pvPortCalloc/pvPortRealloc

**Comment by FishEatingCat** [comment]
> > Normally you compile open62541 as a library so it shouldn't give any error on lacking of other libraries. The header files from LWIP and freeRTOS should be present in your computer, and you should point to them using the right variable in CMake. Both are open source project, so you can just clone the repos.
> > Having them as deps could be too intrusive to this project, IMHO, since the architecture is not on the top list of used architectures.
> 
> As here said, you need to clone the repo of LWIP and of freeRTOS, both are C code. Then in open62541, checkout the master branch, and in CMake, choose freeRTOSLWIP as UA_ARCHITECTURE. You will get then more options. The most important are:
> 
> * UA_ARCH_EXTRA_INCLUDES which you should point to the folders where the headers files from LWIP and freeRTOS, separeted by semicolon (;).
> * UA_ARCH_ADD_FLAGS, UA_ARCH_LINKER_FLAGS and UA_ARCH_REMOVE_FLAGS must be set according to the compiler you're using, so I cannot give more details now here. What I did was: in my IDE where I was compiling my normal C code for freeRTOS, I copy the flags and used them. Then when compilation fails, I had to remove some compilers flags (-Wconversion in my case) which weren't supported or were causing compilation errors.
> * UA_ARCH_FREERTOS_USE_OWN_MEMORY_FUNCTIONS is a boolean that you can set to true if you don't want to use free/malloc/calloc/realloc but the ones from the external implementation of freeRTOS called vPortFree/pvPortMalloc/pvPortCalloc/pvPortRealloc

Thank you very much for your reply!
I have just been working for two months, so what do you mean by UA_ARCH_ADD_FLAGS, UA_ARCH_LINKER_FLAGS and UA_ARCH_REMOVE_FLAGS in the second point you mentioned? Or it is not very clear what should correspond to the compiler, I am using keil 5 to develop the stm32 project.
Thank you

**Comment by cabralfortiss** [comment]
There are flags that your compiler needs to compile to the specific hardware you're using. In my case, for example I was compiling to cortex-m3 core, so I needed "-mcpu=cortex-m3". I never used keil 5 so no clue how to help you there. If you use an IDE which comes with already everything setup, you could look into it. Or when you compile some code, at some point probably yo do "make", so you could do "make VERBOSE=1" and you'll see the detailed compilation, which has the flags, and included paths.

**Comment by FishEatingCat** [comment]
> There are flags that your compiler needs to compile to the specific hardware you're using. In my case, for example I was compiling to cortex-m3 core, so I needed "-mcpu=cortex-m3". I never used keil 5 so no clue how to help you there. If you use an IDE which comes with already everything setup, you could look into it. Or when you compile some code, at some point probably yo do "make", so you could do "make VERBOSE=1" and you'll see the detailed compilation, which has the flags, and included paths.

Thank you very much for your reply!
I will try to set the relevant parameters as you said. I will reply you once there is a message.

**Pull request merged**